### PR TITLE
Laboratory work #3, Vodopyanova Yulia - 21FPL1 

### PR DESCRIPTION
**Pull request by YuliaVodopyanova**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -152,7 +171,12 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        if not all(id in self._id2word for id in encoded_tokens):
+            return None
+        # for id in encoded_tokens:
+        #     if id not in self._id2word:
+        #         return None
+        return tuple(self._id2word.get(token) for token in encoded_tokens)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:179]: you've already checked that all tokens are present in the `_id2word`, why do you still need to use `get`?

**Review by marina-kaz — COMMENTED** [review]
That's a solid start, keep going!

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -173,7 +194,22 @@ def extract_pairs(tokens: tuple[int, ...], window_length: int) -> Optional[tuple
...
+    if not (tokens and isinstance(window_length, int) and window_length >= 2):
+        return None
+    pairs = []
+    for idx, token in enumerate(tokens):
+        for i in range(1, window_length):
+            try:
+                pair = tokens[idx], tokens[idx + i]
+                if pair[0] != pair[1]:
+                    pairs.append(pair)
+            except IndexError:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:206]: if you are aiming at grade higher than 7 I will ask to modify this part of the code

It is wonderful that you are able to catch and process exceptions, yet _passing_ them is considered to be a bad practice as it is virtually ignoring them 

so I would rather you come up with an algorithm that considers those cases

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -233,7 +269,17 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
...
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vrtx in vertex1, vertex2:
+            if vrtx not in self._matrix[0]:
+                self._matrix[0].append(vrtx)
+                self._matrix.append([0 for _ in range(len(self._matrix[0]))])
+                for i in range(1, len(self._matrix[0])):
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:278]: can you explain what this cycle does and how is it different from the one in the 277th line?

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -278,7 +330,15 @@ def calculate_inout_score(self, vertex: int) -> int:
                 number of incidental vertices
         If vertex is not present in the graph, -1 is returned
         """
-        pass
+        if vertex not in self._matrix:
+            return -1
+        count = []
+        for rows in self._matrix:
+            if rows[0] == vertex:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:337]: do all your rows start with a vertex to which they relate?

if so, it was very unclear...

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -278,7 +330,15 @@ def calculate_inout_score(self, vertex: int) -> int:
...
+                for num in rows[self._matrix[0].index(vertex)][1:]:
+                    if num == 1:
+                        count.append(num)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:340]: consider using `sum`

**Review by marina-kaz — COMMENTED** [review]
I appreciate your steady progress, good job! 
Although there are several places where I would like a bit of a clarification

File: lab_3_keywords_textrank/main.py (hunk 3 above)
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:278]: wouldn't `_matrix[len(self._matrix[0])]` have one more zeros than other rows?

Comment (YuliaVodopyanova) [reply lab_3_keywords_textrank/main.py:278]: in line 277: it adds into the list another ones, which contain zeros, their length starts at 1 element - [0] and then with each step, the number of elements increases by one (to the point when it reaches the length)
the one below goes through nested lists and checks if their length equals the length of the first nested list and if it doesn't (only the last one does), it adds zeros

Comment (YuliaVodopyanova) [reply lab_3_keywords_textrank/main.py:278]: well, it is supposed to look like this [[vertex1, vertex2], [0, 0], [0, 0]], i'll try to check

Comment (YuliaVodopyanova) [reply lab_3_keywords_textrank/main.py:278]: my list with verteces doesn't have another 0, should it?

Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py:282]: we've discussed this approach with colleagues and we came to the conclusion that although this approach can be functional it is advised to only use matrix to store 0 or 1, not the actual vertices 

think of a matrix in math: it does not contain names for the rows and columns, it is purely values
please, try to come up with a way to store information about position-vertex links somehow else 

possibly, i'll write a post about it in the channel... as it is a common issue

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -519,7 +590,12 @@ def update_vertex_score(self, vertex: int, incidental_vertices: list[int], score
             scores: dict[int, float]
                 scores of all vertices in the graph
         """
-        pass
+        summa = 0.
+        for vrtx in incidental_vertices:
+            inout_score = 1 / self._graph.calculate_inout_score(vrtx) * scores[vrtx]
+            summa += inout_score
+            weight = summa * self._damping_factor + (1 - self._damping_factor)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:597]: is this the right identation?

**Review by marina-kaz — COMMENTED** [review]
Good job! Keep going :)

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (YuliaVodopyanova) [reply lab_3_keywords_textrank/main.py:597]: i don't understand, i think it's right?
полученные значения просуммировать и умножить на d
далее результат умножения сложить с (1 - d), и результат такого сложения становится новым весом рассматриваемой вершины

**Review by marina-kaz — COMMENTED** [review]
No changes since my last review(

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -385,7 +454,13 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vrtx in vertex1, vertex2:
+            if vrtx not in self._edges.keys():
+                self._edges[vertex1] = self._edges.get(vertex1, []) + [vertex2]
+                self._edges[vertex2] = self._edges.get(vertex2, []) + [vertex1]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:462]: do I understand it correctly: you only add the edge if at lest one vertex was not present in the graph before?

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (marina-kaz) [reply lab_3_keywords_textrank/main.py]: so the ` weight = summa * self._damping_factor + (1 - self._damping_factor)` was supposed to be performed after the cycle has ended?

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -564,7 +662,7 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        return tuple(sorted(sorted(self._scores), key=lambda x: self._scores[x], reverse=True))[:n_keywords]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:665]: Also dictionaries are ordered, it is a byproduct of development, and using it is considered a bad practice

Use `key` argument to establish a secondary sorting order

**Review by marina-kaz — COMMENTED** [review]
Good job!
You can do it for the better score than 6!

File: lab_3_keywords_textrank/main.py [hunk 9]
```diff
@@ -306,11 +290,11 @@ def is_incidental(self, vertex1: int, vertex2: int) -> int:
                 1 if vertices are incidental, otherwise 0
         If either of vertices is not present in the graph, -1 is returned
         """
-        if vertex1 not in self._vertices or vertex2 not in self._vertices:
+        if not (vertex1 in self._vertices and vertex2 in self._vertices):
             return -1
-        idx1 = self._vertices.index(vertex1)
-        idx2 = self._vertices.index(vertex2)
-        return self._matrix[idx1][idx2]
+        if self._matrix[self._vertices.index(vertex1)][self._vertices.index(vertex2)] == 1:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:295]: consider directly returning the value at the intersection of these two vertices: it is already either 0 or 1

File: lab_3_keywords_textrank/main.py [hunk 10]
```diff
@@ -339,8 +323,7 @@ def calculate_inout_score(self, vertex: int) -> int:
         """
         if vertex not in self._vertices:
             return -1
-        idx = self._vertices.index(vertex)
-        return sum(self._matrix[idx])
+        return sum([x[self._vertices.index(vertex)] for x in self._matrix])
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:326]: is there a way to avoid the cycle given the fact that matrix is symmetric?

**Review by marina-kaz — COMMENTED** [review]
Good job, keep going!

**Pull request closed without merging**